### PR TITLE
feat(eigen-client-extra-features): Verifier

### PR DESCRIPTION
**Pull request by gianbelinche**
## What ❔
This PR implements verifier for EigenDA

Since config was changed, recompile zkstack:
`cargo install --path zkstack_cli/crates/zkstack --force --locked`

## Checklist

- [ ] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [ ] Tests for the changes have been added / updated.
- [ ] Documentation comments have been added / updated.
- [ ] Code has been formatted via `zkstack dev fmt` and `zkstack dev lint`.

### CODE REVIEW

File: core/node/da_clients/src/eigen/eigendaservicemanager.rs
Comment (juan518munoz) [inline core/node/da_clients/src/eigen/eigendaservicemanager.rs:1]: This file should go inside of the already existing `generated` directory

**Review by juan518munoz — APPROVED** [review]
LGTM. Be sure to remove all unused imports before merging!

File: core/lib/config/src/configs/da_client/eigen.rs [hunk 1]
```diff
@@ -28,11 +28,13 @@ pub struct DisperserConfig {
     pub eth_confirmation_depth: i32,
     pub eigenda_eth_rpc: String,
     pub eigenda_svc_manager_address: String,
-    pub blob_size_limit: u64,
+    pub blob_size_limit: u32,
```
Comment (juanbono) [inline core/lib/config/src/configs/da_client/eigen.rs:31]: why this change?

Comment (gianbelinche) [reply core/lib/config/src/configs/da_client/eigen.rs:31]: It is needed to call `KZG::setup`, since it receives a u32, and it would not be correct to just downcast it.

File: core/node/da_clients/src/eigen/verifier.rs [hunk 2]
```diff
@@ -0,0 +1,690 @@
...
+#[derive(Debug, Clone)]
+pub struct VerifierConfig {
+    pub verify_certs: bool,
+    pub rpc_url: String,
+    pub svc_manager_addr: String,
+    pub max_blob_size: u32,
+    pub path_to_points: String,
+}
+
+#[derive(Debug, Clone)]
```
Comment (juanbono) [inline core/node/da_clients/src/eigen/verifier.rs:31]: can you add some docs to this struct?

File: core/node/da_clients/src/eigen/verifier.rs [hunk 3]
```diff
@@ -0,0 +1,690 @@
...
+
+        let eigenda_svc_manager = EigenDAServiceManager::new(svc_manager_addr, provider);
+        Ok(Self {
+            kzg,
+            eigenda_svc_manager,
+            cfg,
+        })
+    }
+
+    fn commit(&self, blob: Vec<u8>) -> Result<G1Affine, VerificationError> {
```
Comment (juanbono) [inline core/node/da_clients/src/eigen/verifier.rs:74]: Add docs

File: core/node/da_clients/src/eigen/verifier.rs [hunk 4]
```diff
@@ -0,0 +1,690 @@
...
+
+        for quorum in required_quorums {
+            if !confirmed_quorums.contains_key(&(quorum as u32)) {
+                return Err(VerificationError::WrongProof);
+            }
+        }
+        Ok(())
+    }
+
+    pub async fn verify_certificate(&self, cert: BlobInfo) -> Result<(), VerificationError> {
```
Comment (juanbono) [inline core/node/da_clients/src/eigen/verifier.rs:323]: add docs

**Review by juanbono — CHANGES_REQUESTED** [review]
Add docs

**Comment by gianbelinche** [comment]
> Add docs

Done

File: core/node/da_clients/src/eigen/sdk.rs [hunk 5]
```diff
@@ -63,20 +77,23 @@ impl RawEigenClient {
...
             .await?;
 
-        let verification_proof = blob_info
-            .blob_verification_proof
-            .clone()
-            .ok_or_else(|| anyhow::anyhow!("No blob verification proof in response"))?;
+        let blob_info = blob_info::BlobInfo::try_from(blob_info)
+            .map_err(|e| anyhow::anyhow!("Failed to convert blob info: {}", e))?;
+        self.verifier
+            .verify_commitment(blob_info.blob_header.commitment.clone(), data)
```
Comment (anupsv) [inline core/node/da_clients/src/eigen/sdk.rs:83]: In this code, I was looking at [this file](https://github.com/lambdaclass/zksync-era/blob/main/core/lib/l1_contract_interface/src/i_executor/structures/commit_batch_info.rs#L183). Where does eigenda fit in the `match (self.mode, self.pubdata_da)` result ?

Comment (anupsv) [reply core/node/da_clients/src/eigen/sdk.rs:83]: To clarify, is this computing the commitment locally to see if the DA has computed the same commitment ?

File: core/node/da_clients/src/eigen/verifier.rs [hunk 6]
```diff
@@ -0,0 +1,699 @@
...
+}
+
+impl Verifier {
+    pub fn new(cfg: VerifierConfig) -> Result<Self, VerificationError> {
+        let srs_points_to_load = cfg.max_blob_size / 32;
+        let kzg = Kzg::setup(
+            &format!("{}{}", cfg.path_to_points, "/g1.point"),
+            "",
+            &format!("{}{}", cfg.path_to_points, "/g2.point.powerOf2"),
+            268435456, // 2 ^ 32
```
Comment (anupsv) [inline core/node/da_clients/src/eigen/verifier.rs:51]: small correction, its 2 ^ 28

File: core/node/da_clients/src/eigen/verifier.rs [hunk 7]
```diff
@@ -0,0 +1,699 @@
...
+    pub fn verify_commitment(
+        &self,
+        expected_commitment: G1Commitment,
+        blob: Vec<u8>,
+    ) -> Result<(), VerificationError> {
+        let actual_commitment = self.commit(blob)?;
+        let expected_commitment = G1Affine::new_unchecked(
+            Fq::from(num_bigint::BigUint::from_bytes_be(&expected_commitment.x)),
+            Fq::from(num_bigint::BigUint::from_bytes_be(&expected_commitment.y)),
+        );
```
Comment (anupsv) [inline core/node/da_clients/src/eigen/verifier.rs:95]: We should add a isInSubgroup and IsOnCurve checks.

File: core/node/da_clients/src/eigen/verifier.rs [hunk 8]
```diff
@@ -0,0 +1,699 @@
...
+
+    fn process_inclusion_proof(
+        &self,
+        proof: &[u8],
+        leaf: &[u8],
+        index: u32,
+    ) -> Result<Vec<u8>, VerificationError> {
+        let mut index = index;
+        if proof.len() == 0 || proof.len() % 32 != 0 {
+            return Err(VerificationError::WrongProof);
```
Comment (anupsv) [inline core/node/da_clients/src/eigen/verifier.rs:139]: Would be good to differentiate the errors since the `VerificationError::WrongProof` is used in many places.

File: core/node/da_clients/src/eigen/sdk.rs (hunk 5 above)
Comment (gianbelinche) [reply core/node/da_clients/src/eigen/sdk.rs:83]: We fit in `(L1BatchCommitmentMode::Validium, PubdataSendingMode::Custom)`
The `verifier.verifyCommitment` function is the one that generates a commitment based on the blob and  compares it to the one given by EigenDA.
The code you were looking at is the one that defines what would be sent to the L1, once we make the PR with full M0, that part will send the EigenDA commitment.

**Review by juanbono — APPROVED** [review]
LGTM
